Tag Model

mariajgrimaldi opened the pull request:
This PR adds to eox_tagging plugin:

- Tag model with override `save` method for adding needed validations. Implemented soft delete overriding the delete method in tag.
- Added validator file to validate entries before saving, this file uses validation settings defined as 
**EOX_TAGGING_DEFINITIONS**
- Added a constants file to store constant used in choices fields 
- Added a queryset with queries over the Tag Model. This queryset is used as a manager
- Added test_utils to define fake django models for testing
- Did make upgrade using python 2.7
- Added resource locator field
- Added a general validation method for opaque keys if used in locator_field
- Now owner can be empty
- Added opaque-key package
- Added new model to help tagging (ModelProxy)
- Added modifications to model so it can tag courses
- Added tests for new features. Also, improved docstrings in test module

felipemontoya commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+        """
+        self.__validate_not_update()
+
+        definitions = base_settings.EOX_TAGGING_DEFINITIONS
```
What happens if the definition does not exist?

mariajgrimaldi replied:
Attribute error not handled. Added the setting to common ✓

felipemontoya commented on eox_tagging/settings/common.py:
```diff
@@ -33,6 +43,13 @@
+
 
 def plugin_settings(settings):  # pylint: disable=unused-argument
     """
```
we need to create here an emtpy EOX_TAGGING_DEFINITIONS list

andrey-canon commented on eox_tagging/models.py:
```diff
@@ -0,0 +1,159 @@
+        invalidated_at: date when the tag is soft deleted
+    """
+
+    id = models.AutoField(primary_key=True,)
```
This is the default value, I reckon it is not necessary

andrey-canon replied:
comments ?

mariajgrimaldi replied:
You're right! Sorry for the delay!

andrey-canon commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+        """
+        self.instance = instance
+        self.functions = {  # Functions defined for validation
+            "definition": self.__validate_field_definition
```
trailing comma

andrey-canon commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+            for val in validations:
+                self.functions[val](obj)
+
+    def __validate_not_update(self):
```
if I press save in the admin panel I will get a 500 error ?

mariajgrimaldi replied:
Yes. I think I solved it adding an empty EOX_TAGGING_DEFINITIONS to common

andrey-canon commented on eox_tagging/test/test_models.py:
```diff
@@ -0,0 +1,202 @@
+from eox_tagging.test_utils import CourseFakeModel, EnrollmentsFakeModel
+
+
+@override_settings(EOX_TAGGING_DEFINITIONS=[{"field_name": "tag_value",
```
You can write this in the settings test file

mariajgrimaldi replied:
I did, @felipemontoya told me to change it as a recommendation

andrey-canon replied:
so... @felipemontoya  why here instead of the test settings file ?

felipemontoya replied:
Because we want to change the settings for every test

andrey-canon commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+        definitions = base_settings.EOX_TAGGING_DEFINITIONS
+
+        for obj in definitions:
+            validations = obj["validations"]
```
```suggestion
            validations = obj.get("validations", [])
```

mariajgrimaldi commented:
While trying to change the validations using django api I came across the following problem:

When calling full_clean() django validates I think the integrity of the Foreign Keys, the problem in our case is that we are using Generic Foreign Keys and their behavior is different. For example:

If I have a FK pointing to Users, then full_clean will validate that the Users exists in the table users
When having a GFK (generic foreign key), the GFK will not point to users, instead points Contenttypes. Then, when full_clean validates (using validate unique), raises a validationError because it will not find the User in the Contenttypes table.

What should I do about this? 
I was thinking in using a pre_save signal, it makes more sense to me validating that way.

I think this explains the error I'm having, if not, please correct me

@felipemontoya

andrey-canon commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+        Attributes:
+            obj: the object in definitions used for validations
+        """
+        field = obj["field_name"]
```
use get or catch the exception and handle it

andrey-canon commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+            obj: the object in definitions used for validations
+        """
+        field = obj["field_name"]
+        field_value = getattr(self.instance, field)
```
I think you should handle AttributeError

andrey-canon commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,68 @@
+        """
+        field = obj["field_name"]
+        field_value = getattr(self.instance, field)
+        values_allowed = obj["allowed"]
```
same comments...

felipemontoya commented:
@mariajgrimaldi I think the error is clear, but to solve it what I would do is instead override the method that does the uniqueness validation and have it do the custom logic that we need. That would be validating the existence of said user if it is a user_tag, o the existence of the correct model

andrey-canon commented on eox_tagging/models.py:
```diff
@@ -0,0 +1,151 @@
+
+        return self.valid().filter(
+            belongs_to_type=ctype,
+            belongs_to_object_id=owner.id
```
trailing comma

andrey-canon commented on eox_tagging/models.py:
```diff
@@ -0,0 +1,151 @@
+
+        return self.valid().filter(
+            tagged_type=ctype,
+            tagged_object_id=tagged_object.id
```
trailing comma

andrey-canon commented on eox_tagging/test/test_models.py:
```diff
@@ -0,0 +1,183 @@
+                belongs_to=self.belongs_to_object,
+            )
+
+    def test_tag_different_generic_objects(self):
```
same comment line 46

mariajgrimaldi replied:
What we want to test is we can tag multiple objects, we are testing the generic part of the model. 
The creation must succeed because there is not a restriction in the objects we can tag.
Do you think we can add more to the test?

andrey-canon commented on eox_tagging/test/test_models.py:
```diff
@@ -0,0 +1,183 @@
+        Tag.objects.create(
+            tag_value="anything can be here",
+            tag_type="testType",
+            tagged_object=self.fake_object_tagged_fail,  # In previus tests failed, not in this
```
how do you know that the previous test failed? what test do you refer , line 115?  and typo **previus**

mariajgrimaldi replied:
I'm sorry for the error. I meant this test: test_tag_different_generic_objects_fail

andrey-canon commented on eox_tagging/test/test_models.py:
```diff
@@ -0,0 +1,183 @@
+        {"field_name": "tag_value", "validations": ["definition"], "allowed": r"^a...s$"}])
+    def test_tag_validation_regex_accepts_pattern(self):
+        """ Used to confirm that tags can accept a pattern if defined in settings."""
+        Tag.objects.create(
```
x3

mariajgrimaldi replied:
For example: 
With this setting, the field tag_value has a validation defined, called _definition_. This validation means that the field tag_value must follow the definition described in _allowed_. In this case, it must match the pattern.
What we are testing is that we can create the object that follows that definition. The test _test_tag_validation_regex_accepts_pattern_fail_ demonstrates that if the field does not follow the definition, then the creation fails

mariajgrimaldi commented on eox_tagging/test/test_models.py:
```diff
@@ -0,0 +1,183 @@
+    @override_settings(EOX_TAGGING_DEFINITIONS=[])
+    def test_empty_setting(self):
+        """ Used to test saving without validations defined"""
+        Tag.objects.create(
```
The purpose of the test is create successfully the object without any validation error because there is not any definition set in EOX_TAGGING_DEFINITIONS. That means, that we can create any object without restrictions.
Do you think we can add more to the test?

mariajgrimaldi commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,167 @@
+        """ Function that validates existence of the course."""
+        course_id = getattr(self.instance, object_name).course_id  # Course needs to have course_id
+        try:
+            get_courseware_courses().get_course_by_id(course_id)
```
get_course_by_id returns Http404 from from django.http import Http404, should I only catch that exception?
@andrey-canon

mariajgrimaldi commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,167 @@
+            get_edxapp_user(**data)
+            log.info("EOX_TAGGING  |   Validated user integrity %s", data["username"])
+
+        except Exception:
```
This raises NotFound from rest_framework.exceptions import NotFound. Should I catch that even If I don't yet rest_framework?
@andrey-canon

mariajgrimaldi commented on eox_tagging/validators.py:
```diff
@@ -0,0 +1,295 @@
+                                      .format(field_one=field_tuple[0], field_two=field_tuple[1]))
+
+    # Helper functions
+    def __set_defaults(self):
```
@felipemontoya I didn't implement this because I have a question, where should we call it?

mariajgrimaldi commented:
Hey @felipemontoya ! 
I implemented what we talked about yesterday:
- New validator functions for: in, exist, object, equals, regex
- Config validator
- Also, I changed the function for integrity validator.  I think is better now
- I added some tests for the new functions
What's missing is the implementation for set_default, I left you this [question](https://github.com/eduNEXT/eox-tagging/pull/5/files#r438451820)

Please let me know what you think! :)

mariajgrimaldi commented on README.md:
```diff
@@ -15,7 +15,40 @@ Place your plugin features list.
+  {
+    "tag_type":"example_tag_2",
+    "belongs_to": "Site",
+    "validate_target_type": {"object":"CourseOverview"},
```
Is this possible {"object": ["CourseOverview", "CourseEnrollments"]} right? @felipemontoya

morenol commented on README.md:
```diff
@@ -15,8 +15,123 @@ Place your plugin features list.
 
```
Can we rewrite this in the .rst format?

mariajgrimaldi replied:
I didn't know that! Yes we can :+1:

mariajgrimaldi replied:
Done! What do you think?

morenol commented on requirements/constraints.txt:
```diff
@@ -14,5 +14,7 @@ Django<2.0
 # TODO: Add contraint explanation
 pylint==1.9.3
 pycodestyle==2.5.0
+django-fake-model==0.1.4
```
Can you sort these contraints? And also please add a comment before each constraint in order to identify why that constraint was added

mariajgrimaldi replied:
Yes! I'm on it! Thank you

mariajgrimaldi replied:
Done. Is it better now?

morenol commented on requirements/base.txt:
```diff
@@ -4,5 +4,4 @@
 #
-django==1.11.29           # via -c requirements/constraints.txt, -r requirements/base.in
-pytz==2020.1              # via django
+-e git+https://github.com/edunext/eox-core.git@v2.12.3#egg=eox-core  # via -r requirements/base.in
```
We can use eox-core,now, it is in pypi https://pypi.org/project/eox-core/

mariajgrimaldi replied:
Nice!! Done

morenol commented on requirements/constraints.txt:
```diff
@@ -11,8 +11,11 @@
-pylint==1.9.3
-pycodestyle==2.5.0
+# Working version with Python 2.7 and version used in Edx Platform
+edx-opaque-keys==0.4.4
```
I think that edx-opaque-keys<2.1.0 also supports python2.7 and the master version of edx-platform uses edx-opaque-keys==2.1.0, so I think that if this constraint is needed the comment should be something like Version used in Ironwood version of edx-platform.

Please, also use as constraint something like

edx-opaque-keys<1.0.0

mariajgrimaldi replied:
Done! Thanks for the suggestions

morenol commented on requirements/constraints.txt:
```diff
@@ -11,8 +11,11 @@
\ No newline at end of file
+futures; python_version == "2.7"
+
+# Working version with Python 2.7
```
```suggestion
# pylint version 2.0.0 dropped support for python 2.7
# See: https://github.com/PyCQA/pylint/blob/master/doc/whatsnew/2.0.rst
```

morenol commented on requirements/base.in:
```diff
@@ -1,4 +1,4 @@
 -c constraints.txt
 
-Django
+eox-core
```
Is this plugin installing properly in the devstack without breaking anything? @mariajgrimaldi

mariajgrimaldi replied:
Yes, everything seems to work just fine!

morenol commented on .gitignore:
```diff
@@ -1,6 +1,7 @@
 *.py[cod]
 __pycache__
 .pytest_cache
+db.sqlite3
```
This is added here, but there is a db.sqlite3 file that is going to be created with this PR, is that OK? @mariajgrimaldi
